File: src/go/p1/readers_writers.go
package main

import (
	"fmt"
	"time"
	"sync"
	"math/rand"
	"go/shared/parse"	//Requires that your GOPATH include the directory which contains this repo.
)

func reader(id int, data *int, lock *sync.RWMutex, done chan<- bool) {
	//start := time.Now()
	lock.RLock()
	//totalTime := time.Since(start)
	
	fmt.Printf("(Reader %d) Begins reading...\n", id)
	time.Sleep(time.Duration(rand.Intn(10)) * time.Millisecond)
	value := *data
	value++
	fmt.Printf("(Reader %d) Read %d.\n", id, *data)
	
	lock.RUnlock()
	
	//fmt.Printf("%d\n", totalTime.Nanoseconds())
	
	done <- true
}

func writer(id int, data *int, lock *sync.RWMutex, done chan<- bool) {
	//start := time.Now()
	lock.Lock()
	//totalTime := time.Since(start)
	
	fmt.Printf("(Writer %d) Begins writing...\n", id)
	time.Sleep(time.Duration(rand.Intn(10)) * time.Millisecond)
	*data = *data + 1
	fmt.Printf("(Writer %d) Wrote %d.\n", id, *data)
	
	lock.Unlock()
	
	//fmt.Printf("%d\n", totalTime.Nanoseconds())
	
	done <- true
}

func testScenario(totalReaders, totalWriters int) {
	data := 0
	lock := sync.RWMutex{}
	done := make(chan bool)
	
	for i, j := 0, 0; i < totalReaders || j < totalWriters; {
		time.Sleep(time.Duration(rand.Intn(5)) * time.Millisecond)
		if i < totalReaders && j < totalWriters {
			if rand.Intn(2) == 0 {
				go reader(i, &data, &lock, done)
				i++
			}else{
				go writer(j, &data, &lock, done)
				j++
			}
		}else if i < totalReaders {
			go reader(i, &data, &lock, done)
			i++
		}else if j < totalWriters {
			go writer(j, &data, &lock, done)
			j++
		}
	}
	
	for i := 0; i < totalReaders + totalWriters; i++ {
		<- done
	}
	//fmt.Printf("Final value: %d\n", data)
}

func main() {
	rand.Seed(time.Now().UTC().UnixNano())
	
	fmt.Print("Please input how many reader threads to run: ")
	if readers, readersErr := parse.ScanInt(); !readersErr && readers >= 0 {
		fmt.Print("Please input how many writer threads to run: ")
		if writers, writersErr := parse.ScanInt(); !writersErr && writers >= 0 {
			testScenario(readers, writers)
		}else{
			fmt.Println("Please input a single natural number, and nothing else.")
		}
	}else{
		fmt.Println("Please input a single natural number, and nothing else.")
	}
}

File: src/cpp/p1/readers_writers.cpp
#include <mutex>
#include <thread>
#include <vector>
#include <chrono>
#include <cstdlib>
#include <iostream>
#include <stdexcept>
#include <shared_mutex>
#include "cpp/shared/parse.hpp"

typedef std::chrono::steady_clock testing_clock;

std::mutex output_mutex;	//This protects std::cout while the readers/writers are working.

void reader(int id, int* data, std::shared_mutex* lock){
	//testing_clock::time_point start = testing_clock::now();
	lock->lock_shared();
	//testing_clock::time_point end = testing_clock::now();
	
	output_mutex.lock();
	std::cout << "(Reader " << id << ") Begins reading...\n";
	output_mutex.unlock();
	
	std::this_thread::sleep_for(std::chrono::milliseconds(std::rand() % 10));
	
	int value = *data;
	++value;
	
	output_mutex.lock();
	std::cout << "(Reader " << id << ") Read " << *data << ".\n";
	output_mutex.unlock();
	
	lock->unlock_shared();
	
	/*output_mutex.lock();
	std::cout << std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count() << "\n";
	output_mutex.unlock();*/
}

void writer(int id, int* data, std::shared_mutex* lock){
	//testing_clock::time_point start = testing_clock::now();
	lock->lock();
	//testing_clock::time_point end = testing_clock::now();
	
	output_mutex.lock();
	std::cout << "(Writer " << id << ") Begins writing...\n";
	output_mutex.unlock();
	
	std::this_thread::sleep_for(std::chrono::milliseconds(std::rand() % 10));
	
	*data = *data + 1;
	
	output_mutex.lock();
	std::cout << "(Writer " << id << ") Wrote " << *data << ".\n";
	output_mutex.unlock();
	
	lock->unlock();
	
	/*output_mutex.lock();
	std::cout << std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count() << "\n";
	output_mutex.unlock();*/
}

void test_scenario(int total_readers, int total_writers){
	int data = 0;
	std::shared_mutex lock;
	
	std::vector<std::thread> readers(total_readers);
	std::vector<std::thread> writers(total_writers);
	
	for(int i = 0, j = 0; i < total_readers || j < total_writers;){
		std::this_thread::sleep_for(std::chrono::milliseconds(std::rand() % 5));
		if(i < total_readers && j < total_writers){
			if(std::rand() % 2 == 0){
				readers.push_back(std::thread(reader, i++, &data, &lock));
			}else{
				writers.push_back(std::thread(writer, j++, &data, &lock));
			}
		}else if(i < total_readers){
			readers.push_back(std::thread(reader, i++, &data, &lock));
		}else if(j < total_writers){
			writers.push_back(std::thread(writer, j++, &data, &lock));
		}
	}
	
	for(auto i = readers.begin(); i != readers.end(); ++i){
		if(i->joinable()){
			i->join();
		}
	}
	for(auto i = writers.begin(); i != writers.end(); ++i){
		if(i->joinable()){
			i->join();
		}
	}
	
	//std::cout << "Final value: " << data << "\n";
}

int main(){
	std::srand(std::time(0));
	try{
		std::cout << "Please input how many reader threads to run: ";
		int readers = scan_int();
		if(readers >= 0){
			std::cout << "Please input how many writer threads to run: ";
			int writers = scan_int();
			if(writers >= 0){
				test_scenario(readers, writers);
			}else{
				throw std::invalid_argument("Read a negative value from std::cin.");
			}
		}else{
			throw std::invalid_argument("Read a negative value from std::cin.");
		}
	}catch(const std::invalid_argument& ex){
		std::cout << "Please input a single natural number, and nothing else.\n";
	}
	return 0;
}